### PR TITLE
DCMAW-8229: DynamoDB improvements and OpenAPI spec updates

### PR DESCRIPTION
**Pull request by ahickmann-gds**
​DCMAW-8229

### What changed
- Use `marshall()` method to convert a JS object into  a DynamoDB record
- Assert on DynamoDB command inputs with `aws-sdk-client-mock-jest` (unit tests)
- Update OpenAPI spec `/activeSession`
- Update API tests instructions in the `README` file

Changes deployed to the `dev` environment: 

**BEFORE (`main` branch) - create record without `marshall()`:**
Screenshot 2024-10-13 at 17 03 09

**AFTER (this branch) - create  record with `marshall()`:**
Screenshot 2024-10-13 at 17 03 30

**API tests**
Screenshot 2024-10-13 at 17 28 39

### Why did it change
- Improve DynamoDB unit tests
- To ensure API spec is up to date

## Checklists

- [x] There is a ticket raised for this PR that is present in the branch name
- [ ] No PII data logged. [See guidance here](https://govukverify.atlassian.net/wiki/spaces/DCMAW/pages/3502407722/PII+Logging+Considerations)
- [ ] Demo to a BA, TA, and the team.
- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks

### CODE REVIEW

File: backend-api/src/functions/asyncActiveSession/asyncActiveSessionHandler.ts
Comment (ahickmann-gds) [inline backend-api/src/functions/asyncActiveSession/asyncActiveSessionHandler.ts:1]: It seems that "error" is usually lower case in this repo hence the change below. Also, I thought "Invalid authorization header" might be a better "error_description".

File: backend-api/openApiSpecs/async-public-spec.yaml
Comment (ahickmann-gds) [inline backend-api/openApiSpecs/async-public-spec.yaml:1]: I've mainly reviewed `/activeSession`. I didn't bother looking at the other endpoints in detail - we can do that when we implement them.

File: backend-api/.env.example
Comment (sandymay-dd) [inline backend-api/.env.example:1]: Question (non-blocking): Do you find value in this file? I ask as we provide a script to generate the file for us. 

Wondering if remembering to maintain it is greater than the value it adds.

File: backend-api/README.md [hunk 1]
```diff
@@ -36,14 +36,18 @@ npm run test
...
 2. Deploy your stack
 3. Generate a `.env` file for your deployed stack
-
 ```bash
 # From /backend-api
 sh generate_env_file.sh <stack_name>
 ```
 
-4. Run tests
+4. Add the following environment variables to the `.env` file:
```
Comment (sandymay-dd) [inline backend-api/README.md:44]: Please may we use the generate_env_file.sh for this to make this experience as plug and play as possible :)

File: backend-api/src/functions/services/session/tests/sessionService.test.ts [hunk 2]
```diff
@@ -89,11 +88,29 @@ describe("Session Service", () => {
...
           ],
         });
-
         const result = await sessionService.getActiveSessionId("mockSub");
 
         expect(result.isError).toBe(false);
         expect(result.value).toEqual("mockSessionId");
+        const expectedCommandInput: QueryCommandInput = {
+          ExpressionAttributeValues: {
+            ":currentTimeInSeconds": { N: "1710028800" },
```
Comment (sandymay-dd) [inline backend-api/src/functions/services/session/tests/sessionService.test.ts:97]: Nit: is it possible please to add this value in a comment next to this line?

```jest.useFakeTimers().setSystemTime(new Date("2024-03-10"))```

File: backend-api/package.json [hunk 3]
```diff
@@ -56,6 +56,7 @@
     "@aws-sdk/util-dynamodb": "^3.669.0",
     "@types/node": "^20.16.11",
     "aws-cdk-lib": "^2.161.1",
+    "aws-sdk-client-mock-jest": "^4.0.2",
```
Comment (sandymay-dd) [inline backend-api/package.json:59]: please may we move this to the dev dependencies section :)

**Review by sandymay-dd — APPROVED** [review]

File: backend-api/package.json (hunk 3 above)
Comment (ahickmann-gds) [reply backend-api/package.json]: Done - sorry

File: backend-api/src/functions/services/session/tests/sessionService.test.ts (hunk 2 above)
Comment (ahickmann-gds) [reply backend-api/src/functions/services/session/tests/sessionService.test.ts]: Done, but for all of them!

File: backend-api/.env.example
Comment (ahickmann-gds) [reply backend-api/.env.example:1]: I agree. Should we just delete this file?

Comment (ahickmann-gds) [reply backend-api/.env.example:1]: File deleted

File: backend-api/README.md (hunk 1 above)
Comment (ahickmann-gds) [reply backend-api/README.md]: Done

**Pull request merged**